packages/devtools/lib/src/timeline/frames_bar_chart.dart

```diff
@@ -0,0 +1,136 @@
+    element.style.height = '${chartHeight}px';
+    element.style.paddingTop = '${padding}px';
+    element.style.paddingBottom = '${padding}px';
+
+    timelineController.onFrameAdded.listen((TimelineFrame frame) {
+      final CoreElement frameUI = FrameBar(this, frame);
+      if (element.children.isEmpty) {
+        add(frameUI);
+      } else {
+        if (element.children.length >= maxFrames) {
+          element.children.removeLast();
+        }
+        element.children.insert(0, frameUI.element);
+      }
+    });
+  }
+
+  static const int chartHeight = 200;
+  static const int maxFrames = 120;
+  static const padding = 5;
+
+  FrameBar selectedFrame;
+
+  final StreamController<TimelineFrame> _selectedFrameController =
+      StreamController<TimelineFrame>.broadcast();
+
+  Stream<TimelineFrame> get onSelectedFrame => _selectedFrameController.stream;
+
+  void setSelected(FrameBar frameUI) {
+    if (selectedFrame == frameUI) {
+      return;
+    }
+
+    if (selectedFrame != frameUI) {
+      selectedFrame?.setSelected(false);
+      selectedFrame = frameUI;
+      selectedFrame?.setSelected(true);
+
+      _selectedFrameController.add(selectedFrame?.frame);
+    }
+  }
+}
+
+class FrameBar extends CoreElement {
+  FrameBar(this.framesBarChart, this.frame)
+      : super('div', classes: 'timeline-frame') {
+    layoutVertical();
+
+    drawFrame();
+
+    click(() {
+      framesBarChart.setSelected(this);
+    });
+  }
+
+  // Chart height minus padding on top and bottom.
+  static const maxBarHeight =
+      FramesBarChart.chartHeight - (FramesBarChart.padding * 2);
+
+  // Let a 16ms frame take up 1/3 of the [TimelineFramesUI] height, so we should
+  // be able to fit 48ms (3x16) in [chartHeight] pixels.
+  static const double pxPerMs = maxBarHeight / 48;
+
+  final FramesBarChart framesBarChart;
+  final TimelineFrame frame;
+  CoreElement _cpuBar;
+  CoreElement _gpuBar;
+
+  void drawFrame() {
+    final cpuBarHeight = math.min(maxBarHeight, frame.cpuDurationMs * pxPerMs);
+
+    // If we are going to run out of room to display the frame bar, trim the gpu
+    // portion.
+    final gpuBarHeight =
+        math.min(maxBarHeight - cpuBarHeight, frame.gpuDurationMs * pxPerMs);
+
+    final String cpuTooltip = frame.isCpuSlow
+        ? _slowFrameWarning('GPU', frame.gpuAsMs)
+        : 'GPU: ${frame.gpuAsMs}';
+    final gpuTooltip = frame.isCpuSlow
+        ? _slowFrameWarning('GPU', frame.gpuAsMs)
+        : 'GPU: ${frame.gpuAsMs}';
+
+    _cpuBar = div(c: 'bar bottom')
+      ..element.style.height = '${cpuBarHeight}px'
+      ..element.style.background = _getCpuBarColor()
+      ..element.title = cpuTooltip;
+    _gpuBar = div(c: 'bar top')
+      ..element.style.height = '${gpuBarHeight}px'
+      ..element.style.background = _getGpuBarColor()
+      ..element.title = gpuTooltip;
+
+    element.style.height = '${cpuBarHeight + gpuBarHeight}px';
+
+    add(_gpuBar);
+    add(_cpuBar);
+  }
+
+  String _getCpuBarColor() {
```

Comment:
Keep these methods returning Color objects and then convert at the last minute when setting CSS
that way the conversion to Flutter is easier and we have types instead of Strings in more cases.

Reply:
done.